### PR TITLE
Gp 19492 defer payment start

### PR DESCRIPTION
**Pull request by mflandorfer**

### CODE REVIEW

File: templates/CRM/Contract/Form/PaymentPreview/psp_sepa.tpl [hunk 1]
```diff
@@ -77,22 +79,23 @@
...
+        let date = deferPaymentStart ? new Date(PSP.vars["next_installment_dates"][creditorId]) : new Date();
 
-        const grace = parseInt(CRM.vars["de.systopia.contract/psp_sepa"].grace_days[creditorId]);
-        const notice = parseInt(CRM.vars["de.systopia.contract/psp_sepa"].notice_days[creditorId]);
+        const grace = parseInt(PSP.vars.grace_days[creditorId]);
+        const notice = parseInt(PSP.vars.notice_days[creditorId]);
         const millisecondsInADay = 24 * 60 * 60 * 1000;
 
-        date.setTime(date.getTime() + (notice - grace) * millisecondsInADay);
+        date.setTime(date.getTime() + Math.max(0, (notice - grace)) * millisecondsInADay);
```
Comment (mflandorfer) [inline templates/CRM/Contract/Form/PaymentPreview/psp_sepa.tpl:98]: @pfigel Is it correct to use `Math.max(0, ...)` here to avoid negative values?

Comment (pfigel) [reply templates/CRM/Contract/Form/PaymentPreview/psp_sepa.tpl:98]: Remind me to discuss this in a MidTagCall, this is weird.

**Review by pfigel — CHANGES_REQUESTED** [review]
Found two issues, looks good otherwise. I made a few cosmetic/compat changes.

* When I make a change with the checkbox disabled, the `Defer Payment Start` activity custom field will still be set to "Yes". JSON correctly shows `"defer_payment_start":"0"` and the execution works as expected, so it looks like it's just the custom field.
* There's a small mismatch between the backend implementation and what the frontend shows: for  revives, the backend ignores defer payment start and always behaves like it's set to `0`. We should discuss (maybe in a MidTagCall) whether we want to:
  * Remove the form field for revives and keep existing behaviour
  * Still add the form field, but set it to unchecked by default for revives while giving the user the option to check it. This would require backend changes (in order not to ignore the field for revives) that may be a bit tricky to implement in a backwards-compatible way (we would still want other processes that trigger revives to behave as they do now).

**Review by pfigel — CHANGES_REQUESTED** [review]
I've tested a revive case with the defer option enabled: https://dev2.greenpeace.at/civicrm/contact/view?reset=1&cid=1111910

* Payment preview showed the correct next debit date
* However, the SEPA mandate was created with next collection on `2021-09-03`; the membership was last debited on `2021-08-03`, so this should have been `2022-08-03`

I noticed something else, but this appears to be an existing issue (could also reproduce on master). If you open the revive form for [this](https://dev2.greenpeace.at/civicrm/contact/view?reset=1&cid=2220354) contact, the form initially shows a next debit date of `2021-10-24`. As soon as you make a change e.g. to the amount field or defer checkbox, it switches to `2021-09-03` (which makes a bit more sense).

I can create a separate ticket for that issue if you want, unless it's an easy fix and you want to just include it with this. 😅

**Review by pfigel — APPROVED** [review]
Looks great now, thanks! 🥳

**Pull request merged**